tutorials/language/tensorize.py

```diff
@@ -0,0 +1,286 @@
+# In order to leverage our hypothetical hardware's GEMV instruction,
+# we can tensorize over :code:`j.inner`.
+#
+# Define GEMV Tensorization Intrinsic
+# -----------------------------------
+# Before scheduling the tensorization, we need to first define the intrinsic function for GEMV.
+# It includes two parts, the first is a compute definition of GEMV.
+# TVM uses it to match the computing pattern in the original Matmul schedule.
+# The second is to specify how to execute GEMV on the device,
+# which is done in :code:`intrin_func` below.
+#
+def intrin_gemv(m, l):
+    a = tvm.placeholder((l,), name='a')
+    b = tvm.placeholder((m, l), name='b')
+    k = tvm.reduce_axis((0, l), name='k')
+    c = tvm.compute((m,), lambda i: tvm.sum(a[k] * b[i, k], axis=k), name='c')
+    Ab = tvm.decl_buffer(a.shape, a.dtype,
+                         name="A",
+                         offset_factor=1,
+                         strides=[1])
+    Bb = tvm.decl_buffer(b.shape, b.dtype,
+                         name="B",
+                         offset_factor=1,
+                         strides=[tvm.var("ldw"), 1])
+    Cb = tvm.decl_buffer(c.shape, c.dtype,
+                         name="C",
+                         offset_factor=1,
+                         strides=[1])
+    def intrin_func(ins, outs):
+        ib = tvm.ir_builder.create()
+        aa, bb = ins
+        cc = outs[0]
+        ib.emit(tvm.call_extern("int32", "gemv_update",
+                                cc.access_ptr("w"),
+                                aa.access_ptr("r"),
+                                bb.access_ptr("r"),
+                                m, l, bb.strides[0]))
+        return ib.get()
+    with tvm.build_config(offset_factor=1):
+        return tvm.decl_tensor_intrin(c.op, intrin_func, binds={a: Ab, b: Bb, c: Cb})
+
+######################################################################
+# Here :code:`tvm.decl_tensor_intrin` declares how to execute the computation :code:`c.op`.
+# Our implementation simply takes the inputs and outputs,
+# converts them to pointers and emit an external function call.
+# Note that tensorization requires user to specify :code:`offset_factor`,
+# with this information, TVM has knowledge of whether the data is aligned
+# between the start address of the original data structure
+# and the offset being passed to tensorize,
+# so that it has chance to optimize with vectorized loading.
+# We set the factor to 1 for simplification.
+#
+# Buffers are also declared for inputs and outputs, though this is not required,
+# we benefit from the extra information provided by buffers. For example, we pass
+# :code:`bb.strides[0]` as an argument to the external function :code:`gemv_update`.
+# For now :code:`bb.strides[0] == l`,
+# but later we will see how they can differ with more complicated schedules.
+#
+# Note that we use :code:`tvm.var("ldw")` as the first stride dimension for :code:`B`.
```

Comment:
BTW, I am curious whether 'ldw' has any meaning.

Reply:
I change it to `s1` for stride1 instead.